### ISSUE TITLE
MobX - Do not recommend using at this stage

### ISSUE THREAD

**Issue by ctrlplusb** [ctrlplusb/react-universally#408]
Hey all

MobX does some `componentWillMount` observability bindings that don't play nicely with the current architecture of this starter kit.  For now I recommend against using MobX until I can investigate this further.

💖

**Comment by ctrlplusb** [comment]
Please read #409

**Comment by ctrlplusb** [comment]
@birkir 👆

**Comment by ctrlplusb** [comment]
Hey @birkir and @xiao-hu

I have updated the `feature/mobx` branch (I deleted `feature/mobx@next`) with the latest release of `react-async-bootstrapper` which allows you to set if you would like `componentWillUnmount` to run.  I have only set it to run on the client side bootstrapping but you could experiment with it on the server side too.

Let me know what your experience is like. Thanks!

**Comment by birkir** [comment]
@ctrlplusb what a man. I tried it out and it worked great after I fixed a little bug in passing the options to recursive function. I added a PR for it.

I really appreciate your effort in solving the issue with mobx.

**Comment by ctrlplusb** [comment]
Radness! Have you had luck rehydrating state from server to client?

**Comment by ctrlplusb** [comment]
Version 2.1.1 of `react-tree-walker` released. :)

**Comment by birkir** [comment]
Awesome stuff! About re-hydration, I usually use a single domain store (that contains other stores like UI, User, Items, etc.) that I rehydrate from the server, components then inject the store and I get kind of valid rehydration.

However I sometimes use mobx within a component for state management and haven't got them to re-hydrate (yet...) 👍 

But hey, great work Sean.

**Comment by diondirza** [comment]
@ctrlplusb @birkir I know I can count on you guys. Preparing migration for my production to version 13 I guess.

**Comment by diondirza** [comment]
@ctrlplusb @birkir i have bad message for you guys, this error still happened in feature/mobx branch. Exactly after you refresh on Counter page then click increment/decrement button.

**Comment by ctrlplusb** [comment]
@diondirza can you check your node modules and ensure 2.1.1 is there?

**Comment by diondirza** [comment]
@ctrlplusb just found pulled `react-tree-walker` in `node_modules` still 2.1.0 sean, probably you use exact version on its dependency.

**Comment by ctrlplusb** [comment]
run a `yarn upgrade` for now though 😊

**Comment by diondirza** [comment]
The problem caused by `yarn.lock` file here. Now it works fine. gr8

**Comment by diondirza** [comment]
Anyways, I make a little experiment here. Try to call  `counter.inc` function in componentWillMount, the server rendered result different with client rendered result

<img width="1429" alt="screen shot 2017-04-06 at 10 49 48 pm" src="https://cloud.githubusercontent.com/assets/10954870/24763446/a17621f2-1b1b-11e7-9e4d-5459554a046c.png">

Executed 2 times in server and 2 times in client, so I guess for now better using componentDidMount event to have same value both client and server.

**Comment by ctrlplusb** [comment]
@diondirza if you are going to do state modification on server you need to send that state back to the client somehow so that it can use it to start with.  I know how to do this with `redux`, but have no `mobx` experience.

**Comment by diondirza** [comment]
Actually my concern is not maintain modified state to have same value both client and server. Instead I usually set flag in my store on componentWillMount and that flag has side-effect to do ajax call to fetch some data. Probably need to re-tweak some of my code for this changes.

**Comment by ctrlplusb** [comment]
Yeah, you could try use the `asyncBootstrap` function that is powered by `react-async-bootstrapper` for this type of thing. Any component that has this will have it executed during the bootstrap phase.  Just return a promise that only resolves after your ajax is complete and you have updated your mobx state. 👍

**Comment by diondirza** [comment]
Cool, just look at it's README, gonna continue to try it tomorrow. It's pretty late here. Anyway good job guys.

**Comment by xiao-hu** [comment]
@ctrlplusb, I upgraded to the lastest version of `react-async-bootstrapper` and added `{componentWillUnmount: true}` to the function `asyncBootstrapper`. It worked great!!!

**Comment by mattiasewers** [comment]
Worked for me too
